### DISCUSSION TITLE
Accepting service/category without open preferences and updating without reload

### DISCUSSION THREAD

**Discussion by arthurliebhardt** [orestbida/cookieconsent#471] [Q&A]
Hello,

is there a good way to accept a category via a button like `cc.acceptCategory('analytics')` and get the updated result in the [getUserPreferences](https://cookieconsent.orestbida.com/reference/api-reference.html#getuserpreferences) immediate?
Or do I need to work with the localstorage here, for example? Somehow struggling getting the updated state after pushing the button without reloading the site.

Thank you and best,

**Comment by orestbida** [comment]
I'm not sure I understand what the issue is.

Adding an event listener to the button's click event should do the job. Why do you need localstorage?

**Reply by arthurliebhardt** [reply]
> I'm not sure I understand what the issue is.
> 
> Adding an event listener to the button's click event should do the job. Why do you need localstorage?

My use case is, I have a contact form which uses reCAPTCHA. The cookies of reCAPTCha are for example in the 'functional' category. When the user goes to the contact form and don't have the 'functional' category accepted I want to show a button, 'Accept Functional Cookies' and then show the contact form without refreshing the site.

Im using next.js and checking with `{cc.getUserPreferences().acceptedCategories.includes('functional')` if the user accepted the functional category or not.

If not accepted I show the user a button  <button type='button' onClick={() => acceptFunctional()}>Accept Functional Cookies</button></>` and the `acceptFunctional()` has following code ```cc.acceptCategory('functional')```

**Reply by orestbida** [reply]
I think I got it (maybe). Here is how I'd achieve this in next.js: https://stackblitz.com/edit/nextjs-s7aw3r?file=pages%2F_app.js

**Reply by arthurliebhardt** [reply]
Exactly, working like I expected it. Thank you very much.

I somehow was not aware that I could do. Either gave `CookieConsentApi` into the component or created a context.
```
import {
    acceptService,
    getUserPreferences,
    acceptCategory,
} from 'vanilla-cookieconsent';
```

**Reply by orestbida** [reply]
I think the above example (using context) is overkill for this usecase.

Here is a much simpler and shorter [example with props](https://stackblitz.com/edit/nextjs-3vdlmr?file=pages%2Findex.js,pages%2F_app.js).

**Reply by arthurliebhardt** [reply]
Ok, will try the other approach, thank you.

when I have the service defined

``` functional: {
      enabled: false,
      readOnly: false,
      services: {
        recaptcha: {
          label: 'Google reCAPTCHA',
          onAccept: () => {
            console.log("enabled recaptcha");
          },
          onReject: () => {
            console.log("disabled recaptcha");
          }
        },
      },
      autoClear: {
        cookies: [
          {
            name: /^(_ga|_gid)/,
          },
        ],
      },
    },
  },
  ```
  
  How do I describe best only the cookies for this service under or is this not possible?
  
  ```
sections: [
            {
              title: 'Cookie Usage',
              description:
                'I use cookies to ensure the basic functionalities of the website and to enhance your online experience. You can choose for each category to opt-in/out whenever you want. For more details relative to cookies and other sensitive data, please read the full <a href="#" class="cc__link">privacy policy</a>.',
            },
            {
![Screenshot 2023-02-13 at 13 09 07](https://user-images.githubusercontent.com/4281468/218454020-91106160-f3d2-44ad-b300-8c4fd480a45a.png)

              title: 'Essential <span class="pm__badge">always enabled</span>',
              description: 'Description',
              linkedCategory: 'necessary',
              cookieTable: {
                headers: {
                  name: 'Name',
                  domain: 'Service',
                  description: 'Description',
                  expiration: 'Expiration',
                },
                body: [
                  {
                    name: 'cc_cookie',
                    domain: 'domain.com',
                    description: 'Cookie set by cookieconsent plugin <a href="https://cookieconsent.orestbida.com/">CookieConsent v3</a>',
                    expiration: 'Expires after 182 days',
                  }
                ],
              }
            },
            {
              title: 'Functional',
              linkedCategory: 'functional',
              cookieTable: {
                headers: {
                  name: 'Name',
                  domain: 'Service',
                  description: 'Description',
                  expiration: 'Expiration',
                },
                body: [
                  {
                    name: '_ga',
                    domain: 'Google Analytics',
                    description:
                      'Cookie set by <a href="#das">Google Analytics</a>.',
                    expiration: 'Expires after 12 days',
                  },
                  {
                    name: '_gid',
                    domain: 'Google Analytics',
                    description:
                      'Cookie set by <a href="#das">Google Analytics</a>',
                    expiration: 'Session',
                  },
                ],
              },
            },
            {
              title: 'More information',
              description:
                'For any queries in relation to my policy on cookies and your choices, please <a class="cc__link" href="#yourdomain.com">contact me</a>.',
            },
          ],
        },
```
Currently the preferences look like in the attached screenshot
![Screenshot 2023-02-13 at 13 09 07](https://user-images.githubusercontent.com/4281468/218454107-66178eba-78a7-4f69-bc78-7ea29f47e06d.png)

**Reply by orestbida** [reply]
Unfortunately it's not possible to list cookies under a service, and probably will stay like this. Services were added as an extra, for a more granular control (to not force the user into accepting the entire category). Most don't even use services, that's why I decided to keep just one cookie table per category.

**Reply by arthurliebhardt** [reply]
Oh ok, make sense. So if I only need the functional cookies I can also leave the service stuff away but don't have the onAccept / onReject methods right? Or how would the demo above with out services would look like?

**Reply by orestbida** [reply]
[Here you go](https://stackblitz.com/edit/nextjs-ns1hyp?file=pages%2Findex.js,components%2FCookieConsentComponent.js).

**Reply by arthurliebhardt** [reply]
Thank you, that helps a lot.
